### DISCUSSION TITLE
How to define a dependency relationship between a multipartitioned asset and a single dimension partitioned asset

### DISCUSSION THREAD

**Discussion by clairelin135** [dagster-io/dagster#15675] [Q&A] [answered]
Say I have a multipartitioned asset with two dimensions:
- projects (dynamic partition)
- weeks (time partition)

I want to create a downstream asset that is partitioned only on the weeks dimension, so that it can aggregate across projects. Is this possible?

**Comment by clairelin135** [comment, marked as answer]
Yes, this is possible.

When Dagster detects a dependency relationship between a multipartitioned asset and a single-dimension partitioned asset, **where the single dimension partitions def is a dimension of the multipartitions def**, Dagster automatically uses the `MultiToSingleDimensionPartitionMapping`. For a partition key X, this partition mapping assumes that any multi-partition key with X in the matching dimension is a dependency.

In the example below, lets say the projects partitions def has partitions `a`, `b`, and `c`:
```python
from dagster import (
    DynamicPartitionsDefinition,
    WeeklyPartitionsDefinition,
    asset,
    MultiPartitionsDefinition,
)

projects_partitions_def = DynamicPartitionsDefinition(name="project") # a, b, c


@asset(
    partitions_def=MultiPartitionsDefinition(
        partitions_defs={
            "week": WeeklyPartitionsDefinition("2023-01-01"),
            "project": projects_partitions_def,
        }
    )
)
def upstream():
    ...


@asset(partitions_def=projects_partitions_def)
def downstream(context, upstream):
    ...
```

When you execute `downstream` with partition `a`, it will load every partition of `upstream` that has `a` as its `project` dimension: `a|2023-01-01`, `a|2023-01-08`, etc.

**Reply by thegartrell** [reply]
That's great, makes sense and is really convenient. 

But I'm still unclear on how I access the output of `upstream`. In my case, when I write a downstream asset with `context` and `upstream`, the materialization throws a KeyError:

```
dagster._core.errors.DagsterExecutionLoadInputError: Error occurred while loading input "upstream" of step "downstream":
The above exception was caused by the following exception:
KeyError: 'a|2023-06-18T00:00:00Z'
```

Do I still need to define a partition mapping in the @asset decorator?

**Reply by clairelin135** [reply]
@thegartrell You do not need to define a partition mapping, the partition mapping is already inferred to be the `MultiToSingleDimensionPartitionMapping`.

It looks like an error is occurring within the `load_input` call of your IO manager when attempting to load `upstream`. Which IO manager is the upstream asset using? The partition key being fetched is also formatted incorrectly, which I think may be related to the error you're getting--weekly partitions defs should have keys like `2023-06-18`, not `2023-06-18T00:00:00Z`

**Reply by thegartrell** [reply]
I'm using the s3_pickle_io_manager

**Reply by clairelin135** [reply]
Are you on 1.4.3 or earlier? The `KeyError` bug was fixed in 1.4.4: https://github.com/dagster-io/dagster/pull/15528

**Reply by thegartrell** [reply]
I was sourcing the wrong venv which was using 1.3.10. The key error is not appearing anymore and the aggregation works. Thank you!

**Reply by mikolak-net** [reply]
@clairelin135 : Just a short follow-up question – how does one define a job from this kind of asset sequence? Regardless of what was being attempted (e.g. `define_asset_job`), the following was always emitted:

```
dagster._core.errors.DagsterInvalidDefinitionError: Multiple partitioned assets exist in assets job '<job_name>'. Selected assets must have the same partitions definitions, but the selected assets have different partitions definitions:
Multi-partitioned, with dimensions:
Batch_id: Dynamic partitions: "<partition_key_1>"
Video_ids: Dynamic partitions: "<partition_key_2>": {AssetKey(['upstream'])}
Dynamic partitions: "<partition_key_1>": {AssetKey(['downstream'])}
```

It looks like this a standard check for a *job* (not *asset*) creation, present in even the latest version. And, unfortunately, a job definition is – to my understanding – required to specify something like a sensor.

**Reply by clairelin135** [reply]
@mikolak-net Unfortunately defining a job that contains assets with different partitions defs is not currently supported, because there currently is a constraint that each run can only execute on the same partition for each asset.

Not sure what your use case is for defining a sensor, but you could have two sensors -- one for the multipartitioned asset, one for the dynamic asset. Each sensor can track when its respective asset has been materialized, and kick off a run of the downstream asset(s).

**Reply by mikolak-net** [reply]
@clairelin135 : thanks for the answer. Fingers crossed that constraint gets relaxed eventually.

As to "why a sensor" – good question! The specific scenario is a bit different then the original example: *both* constituent partitions are dynamic, and their keys are determined based on an external data source, whose change should also trigger materialization. Hence the need for a sensor.

**Reply by achen187** [reply]
The constraint in a job makes sense, but I think in an ideal world, the job would be smart enough to handle this situation. It could detect "hey theres different partition keys, but one is just a single dimension version of the multi dimension", which then just uses the MultiToSingleDimensionPartitionMapping by default.

We have a use case where this would be helpful, where we would decrease the dimensions of our assets from 2 dimensions to 1 dimension over the dependency tree. We can get around it with the double sensor solution though, so that's helpful!
